### ISSUE TITLE
SvelteKit: 'SyntaxError: Cannot use import statement outside a module'

### ISSUE THREAD

**Issue by RobinBomkamp** [module-federation/vite#163]
Hey there,

we are having an issue with the version `1.1.3` of this package and a SvelteKit application.

## Works with previous version: 

Yes, it used to work with the version `0.2.8`.

## Steps to reproduce:

1. Create an host with `npm create svelte@latest mf`
2. Create a 'Button'-component.  Button.svelte: `<button>Test</button>`
3. `npm add @module-federation/vite --save-dev`  
4. Add module federation config plugin like below
```
export default defineConfig({
	plugins: [
		sveltekit(),
		federation({ 
      			name: "remote",
      			filename: "remoteEntry.js",
      			exposes: {
        			"./remote-app": "./src/lib/Button.svelte",
      			},
    		}),
	]
});
```
5. Run `npm run build`

## Error: 

Now we are experiencing the following error. 

```
> test@0.0.1 build
> vite build

vite v5.4.9 building SSR bundle for production...
✓ 95 modules transformed.
vite v5.4.9 building for production...
node_modules/@module-federation/sdk/dist/index.cjs.js (636:77): Use of eval in "node_modules/@module-federation/sdk/dist/index.cjs.js" is strongly discouraged as it poses security risks and may cause issues with minification.
✓ 84 modules transformed.
.svelte-kit/output/client/_app/version.json                                  0.03 kB │ gzip:  0.05 kB
.svelte-kit/output/client/.vite/manifest.json                                3.62 kB │ gzip:  0.66 kB
.svelte-kit/output/client/_app/immutable/chunks/hostInit.DIJ0Opoq.js         0.05 kB │ gzip:  0.07 kB
.svelte-kit/output/client/_app/immutable/entry/start.DXYuzCOz.js             0.14 kB │ gzip:  0.12 kB
.svelte-kit/output/client/_app/immutable/chunks/Button.BfsWK_OT.js           0.41 kB │ gzip:  0.31 kB
.svelte-kit/output/client/_app/immutable/chunks/virtualExposes.7lQyljI4.js   0.45 kB │ gzip:  0.33 kB
.svelte-kit/output/client/_app/immutable/nodes/0.V07vbrzA.js                 0.60 kB │ gzip:  0.39 kB
.svelte-kit/output/client/_app/immutable/nodes/2.DYBzjjm4.js                 0.69 kB │ gzip:  0.45 kB
.svelte-kit/output/client/_app/immutable/nodes/1.BWVLcEKo.js                 1.02 kB │ gzip:  0.60 kB
.svelte-kit/output/client/_app/immutable/chunks/preload-helper.C1FmrZbK.js   1.14 kB │ gzip:  0.68 kB
.svelte-kit/output/client/_app/immutable/chunks/scheduler.BvLojk_z.js        2.16 kB │ gzip:  1.02 kB
.svelte-kit/output/client/_app/immutable/entry/app.B2gDDm4Z.js               5.02 kB │ gzip:  1.93 kB
.svelte-kit/output/client/_app/immutable/chunks/index.DFTQtrJW.js            5.43 kB │ gzip:  2.30 kB
.svelte-kit/output/client/_app/immutable/chunks/client.CKQRfb6L.js          28.64 kB │ gzip: 11.30 kB
.svelte-kit/output/client/remoteEntry.js                                    66.34 kB │ gzip: 19.36 kB
✓ built in 845ms
(node:48812) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)

node:internal/event_target:1094
  process.nextTick(() => { throw err; });
                           ^
C:\Users\rbom\Desktop\Test\svelte-test\test\node_modules\@module-federation\sdk\dist\index.esm.js:1
import { _ as _extends } from './polyfills.esm.js';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at wrapSafe (node:internal/modules/cjs/loader:1378:20)
    at Module._compile (node:internal/modules/cjs/loader:1428:41)
    at Object..js (node:internal/modules/cjs/loader:1548:10)
    at Module.load (node:internal/modules/cjs/loader:1288:32)
    at Function._load (node:internal/modules/cjs/loader:1104:12)
    at cjsLoader (node:internal/modules/esm/translators:346:17)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:286:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:234:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:473:24)
    at async prerender (file:///C:/Users/rbom/Desktop/Test/svelte-test/test/node_modules/@sveltejs/kit/src/core/postbuild/prerender.js:42:21)
Emitted 'error' event on Worker instance at:
    at [kOnErrorMessage] (node:internal/worker:326:10)
    at [kOnMessage] (node:internal/worker:337:37)
    at MessagePort.<anonymous> (node:internal/worker:232:57)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:820:20)
    at MessagePort.<anonymous> (node:internal/per_context/messageport:23:28)

Node.js v20.18.0
```

## Addition:

Even changing the module federation config to following produces the error:

```
export default defineConfig({
	plugins: [
		sveltekit(),
		federation({ 
     			 name: "remote",
    		}),
	]
});
```

## Our guess

It has something to do with SSR.

**Comment by bastianschaffeld** [comment]
I'm encountering the same issue with my SvelteKit projekt. It looks like it tries to use `require` server side which fails because it's not a browser environment

**Comment by gioboa** [comment]
@RobinBomkamp ~~can you share a repo with a basic example to reproduce the issue please?~~

----

I saw the steps, thanks

**Comment by RobinBomkamp** [comment]
I have build a simple Stackblitz example: https://stackblitz.com/edit/vitejs-vite-wpl99h?file=vite.config.ts&terminal=dev

You have to quit the dev build in the terminal and execute `npm start`.

**Comment by gioboa** [comment]
Thanks @RobinBomkamp I'm glad you have a fallback with v0.2.8, I'll check how to fix the new version in these days.

**Comment by Thiedze** [comment]
I run into the same error.

**Comment by gioboa** [comment]
@Thiedze are you using SvelteKit too?

**Comment by Thiedze** [comment]
Yes @gioboa

**Comment by gioboa** [comment]
I created [this example](https://github.com/gioboa/sveltekit-microfrontend-demo)
It's working in preview mode and with the [flag browser](https://kit.svelte.dev/docs/faq#how-do-i-use-x-with-sveltekit-how-do-i-use-a-client-side-only-library-that-depends-on-document-or-window)

**Comment by RobinBomkamp** [comment]
I tested your example.

It seems like the remote is working with `pnpm build && pnpm preview` as well as `pnpm dev`.
Moreover, the remote exposes a remoteEntry.js in both ways. If you add `manifest: true` to the config. You can see that the only difference is the `publicPath`. (Dev: `publicPath: auto` | Build & preview: `publicPath: './'`)

Left: preview | Right: dev
![image](https://github.com/user-attachments/assets/e4678309-2447-4d1d-ab26-5d070cfe437f)

Both seems to work with a host in `preview`.

This shows me that the problem is part of the host.

If you compare the `manifest.json` of `preview` and `dev`. You can see that the remote is not part of the configuration of the host:

Left: preview | Right: dev
![image](https://github.com/user-attachments/assets/db22798e-0150-4bc8-9ddf-a9cdd48b0e84)

At the moment I don't know why it does not get filled.

**Comment by gioboa** [comment]
We are loading the MFE at runtime though, on a SSR architecture I think it's better to load them on the server.

**Comment by RobinBomkamp** [comment]
Yeah, the remotes should be loaded on the server. I totally agree. 
Is there a way to accomplish this?

The Module Federation Roadmap says that it is planned for the future. Is this already possible now?
https://module-federation.io/guide/start/index.html#%EF%B8%8F-the-future-of-module-federation

**Comment by gioboa** [comment]
we are using module federation under the hood so we need to wait.
I tried the SvelteKit POC on my Ubuntu and Mac but the dev mode is not working. :thinking:

**Comment by RobinBomkamp** [comment]
Do you know if there is a release date or any other plans for module federation with SSR?

**Comment by gioboa** [comment]
> Do you know if there is a release date or any other plans for module federation with SSR?

This a great question but I have no clue about it. cc @ScriptedAlchemy 

> It seems like the remote is working with pnpm build && pnpm preview as well as pnpm dev.

Is SvelteKit POC `dev mode` working for you?

**Comment by RobinBomkamp** [comment]
By 'SvelteKit POC' do you mean your example?

If yes: No, it does not work for me. It seems like it could not find the import statement during compilation.

But I don't understand why `pnpm build` is working and `pnpm dev` not. If so, shouldn't both fail?

**Comment by RobinBomkamp** [comment]
Oh sorry, I have misread your second question.

If I start the remote with `pnpm dev` and the host with `pnpm build && pnpm preview`. It is working out of the box.

**Comment by gioboa** [comment]
Yep, the problems seems host with `dev` mode, it's not resolving the loadRemote promise

**Comment by ScriptedAlchemy** [comment]
> Do you know if there is a release date or any other plans for module federation with SSR?

Federation should work fine on node targets and ssr. Not sure about vite however, I don't think vite has a webpack runtime equivalent for chunk handlers?

**Comment by ScriptedAlchemy** [comment]
You may be able to use something like module loaders. But memory leaks will be an issue with esm. Ideally cjs output is best for servers as we can clear require cache. Esm cache is not accessible from js side.

**Comment by RobinBomkamp** [comment]
> Federation should work fine on node targets and ssr. Not sure about vite however, I don't think vite has a webpack runtime equivalent for chunk handlers?

Sounds amazing. @ScriptedAlchemy Do you have any example on SSR with webpack?

**Comment by RobinBomkamp** [comment]
@gioboa Do you have an idea how we can achieve ssr and module federation with vite?

**Comment by gioboa** [comment]
We are focusing on different issues / features at the moment. Any contributions are more than welcome though

**Comment by RobinBomkamp** [comment]
> We are focusing on different issues / features at the moment. Any contributions are more than welcome though

I can totally understand that. Do you have any information for me to contribute to this project? Maybe I'll find the time.

Even if ssr is supported by module federation it should not break the vite build.

**Comment by ScriptedAlchemy** [comment]
You can look at the node or ssr examples in examples repo. 
We have for modernjs meta framework as well. Next.js too but I no longer maintain next integration

**Comment by jacobtucker-8451** [comment]
Was there a resolution to this? I'm also running into this issue and reproduced it with this simple [StackBlitz](https://stackblitz.com/edit/sveltejs-kit-template-default-bwmaoa?file=vite.config.js). `npm run build` fails with the error at the top of this thread. All I did was generate a SvelteKit project, `npm install --save @module-federation/vite`, and add the minimal module-federation plugin config to `vite.config.js`.

Furthermore, running the called out dependency @module-federation/sdk (where the dependency chain is vite -> runtime -> sdk) through [publint](https://publint.dev/@module-federation/sdk@0.6.14) shows that the index.esm.js filename is problematic since `"type": "module"` is not specified. Is there a trick to get the library to build?

**Comment by gioboa** [comment]
@ScriptedAlchemy do you have any insights for that?

**Comment by ScriptedAlchemy** [comment]
> Was there a resolution to this? I'm also running into this issue and reproduced it with this simple [StackBlitz](https://stackblitz.com/edit/sveltejs-kit-template-default-bwmaoa?file=vite.config.js). `npm run build` fails with the error at the top of this thread. All I did was generate a SvelteKit project, `npm install --save @module-federation/vite`, and add the minimal module-federation plugin config to `vite.config.js`.
> 
> Furthermore, running the called out dependency @module-federation/sdk (where the dependency chain is vite -> runtime -> sdk) through [publint](https://publint.dev/@module-federation/sdk@0.6.14) shows that the index.esm.js filename is problematic since `"type": "module"` is not specified. Is there a trick to get the library to build?

Type module isn't needed if package exports exists afik

**Comment by jacobtucker-8451** [comment]
WIthout it, it defaults to CommonJS, but correctly targets the index.esm.js due to the exports. So even though that's written with ESM syntax, the build sees it as CommonJS since it doesn't end in .mjs.  I'm still learning all the implications of ESM vs CJS but that's what I've found so far.

**Comment by veeramarni** [comment]
For SSR nodejs, [pure ESM is needed](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c) unless wrapped with some vite plugin which is not an ideal case.

**Comment by ScriptedAlchemy** [comment]
So what's the solution for shipping dual standard packages?

Make esm emit with mjs extentions?

**Comment by jacobtucker-8451** [comment]
> So what's the solution for shipping dual standard packages?
> 
> Make esm emit with mjs extentions?

I believe so. That would let consumers interpret it as ESM. [publint](https://publint.dev/rules#implicit_index_js_invalid_format) shows the rules used for that resolution. Under the current case, that index file is interpreted as CJS due to the second bullet.

**Comment by ScriptedAlchemy** [comment]
hows this?
https://github.com/module-federation/core/pull/3159

**Comment by veeramarni** [comment]
The changes look good please release once it is released we can test it.

**Comment by ScriptedAlchemy** [comment]
merged and released

**Comment by RobinBomkamp** [comment]
Is there a way to tell @module-federation/vite to use version 0.7.0 of @module-federation/runtime? Currently it is referenced like this "^0.6.8".

**Comment by gioboa** [comment]
This question is related to https://github.com/module-federation/vite/issues/157 
you can clone the repository and compile a new version locally with @module-federation/runtime -> 0.7.0
if it will solve the issue I'll cut a new release.

**Comment by ScriptedAlchemy** [comment]
Might be able to override with pnpm

**Comment by gioboa** [comment]
> Might be able to override with pnpm 

Here is the [pnpm docs](https://pnpm.io/package_json#pnpmoverrides) with the override explanation.

**Comment by RobinBomkamp** [comment]
It seems like the version 0.7.0 of @module-federation/runtime is not working for this problem. 

I have changed the dependency in a local variant of the @module-federation/vite package and used it in this https://github.com/gioboa/sveltekit-microfrontend-demo repo.

**Comment by ScriptedAlchemy** [comment]
edit node modules manually, then send a PR with whatever needs to be fixed in the core repo

**Comment by zhangHongEn** [comment]
```js
(config.resolve as any).alias.push({
  find: '@module-federation/runtime',
  replacement: require.resolve('@module-federation/runtime'),
});
```
Yes, that’s exactly why ESM cannot be used. This issue alone is easy to fix, but I’m not sure if there are other underlying problems.

**Comment by zhangHongEn** [comment]
https://module-federation.io/configure/implementation.html

If you’re willing, you can also submit a PR for this enhancement, which will resolve the issue.

**Comment by RobinBomkamp** [comment]
> ```js
> (config.resolve as any).alias.push({
>   find: '@module-federation/runtime',
>   replacement: require.resolve('@module-federation/runtime'),
> });
> ```
> 
> Yes, that’s exactly why ESM cannot be used. This issue alone is easy to fix, but I’m not sure if there are other underlying problems.

Do you know a way to test the compatibility? Or do you have a migration guide?

**Comment by zhangHongEn** [comment]
set implementation

---- Replied Message ----
| From | Robin ***@***.***> |
| Date | 11/25/2024 20:14 |
| To | module-federation/vite ***@***.***> |
| Cc | zhn ***@***.***>,
Comment ***@***.***> |
| Subject | Re: [module-federation/vite] SvelteKit: 'SyntaxError: Cannot use import statement outside a module' (Issue #163) |

(config.resolveasany).alias.push({find: ***@***.***/runtime',replacement: ***@***.***/runtime'),});

Yes, that’s exactly why ESM cannot be used. This issue alone is easy to fix, but I’m not sure if there are other underlying problems.

Do you know a way to test the compatibility? Or do you have a migration guide?

**Comment by RobinBomkamp** [comment]
Yeah, but I've meant for the underlying problems.

**Comment by RobinBomkamp** [comment]
I have made some changes in my repo. Have a look here: https://github.com/RobinBomkamp/vite-inject-entry/tree/svelte-kit

I am open to create a PR, but first this needs to be merged: https://github.com/module-federation/vite/pull/193

**Comment by gioboa** [comment]
> I am open to create a PR, but first this needs to be merged: #193

Merged. 1.1.8 is out with your improvement.
